Surface Interpolation Function [Fixes Issue45]

ojeda-e opened the pull request:
This is a starting point to solve #45 

Changes in this PR:

- [x] Added function to calculate interpolation of z_surface (`surface_interpolation`). Includes docstrings.
- [x] Added tests of `surface_interpolation`. 

As suggested by @orbeckst in this [comment](https://github.com/MDAnalysis/membrane-curvature/issues/45#issuecomment-881972548), the function here included uses NumPy. Thanks for encouraging numpy! 

Surfaces included in tests are dummy_arrays with the following characteristics:
- size 3x3 with all values 150 and one `np.nan`.
- size 4x3 with all values 150 and two `np.nan`.
- size 4x4 with all values 150 and two `np.nan`.
- size 3x3 array 3x3  with lots of `np.nan`. 💪🏽 

Probably I'll be asked to add more tests, but this is ok to start. Probably @lilyminium will give me lots of ideas. 😅 

 Thanks.

Commit:
Added interpolation function with docstrings

Commit:
Added surface_interpolation tests

pep8speaks commented:
Hello @ojeda-e! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`membrane_curvature/surface.py`](https://github.com/MDAnalysis/membrane-curvature/blob/8206fdadacfb31a9af40b2b9cc94fc9ba94218e0/membrane_curvature/surface.py):

> [Line 144:33](https://github.com/MDAnalysis/membrane-curvature/blob/8206fdadacfb31a9af40b2b9cc94fc9ba94218e0/membrane_curvature/surface.py#L144): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 156:61](https://github.com/MDAnalysis/membrane-curvature/blob/8206fdadacfb31a9af40b2b9cc94fc9ba94218e0/membrane_curvature/surface.py#L156): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`membrane_curvature/tests/test_membrane_curvature.py`](https://github.com/MDAnalysis/membrane-curvature/blob/8206fdadacfb31a9af40b2b9cc94fc9ba94218e0/membrane_curvature/tests/test_membrane_curvature.py):

> [Line 6:120](https://github.com/MDAnalysis/membrane-curvature/blob/8206fdadacfb31a9af40b2b9cc94fc9ba94218e0/membrane_curvature/tests/test_membrane_curvature.py#L6): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 119 characters)

Commit:
Fixed PEP8

Commit:
Fixed PEP8

ojeda-e commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -208,6 +209,39 @@ def test_get_z_surface(x_bin, y_bin, x_range, y_range, expected_surface):
+    (np.array([[150.,  150, 150.,  150.],
+               [150., np.nan, np.nan,  150.],
+               [150., 150., 150.,  150.]]),
+     np.array([[150., 150,  150.,  150.],
```
I just noticed I didn't use `np.full` here. Booo.

tylerjereddy commented on membrane_curvature/surface.py:
```diff
@@ -97,11 +97,68 @@ def normalized_grid(grid_z_coordinates, grid_norm_unit):
+    ----------
+
+    array_surface: np.ndarray
+        Numpy array of floats of shape (`n_x_bins`, `n_y_bins`)
```
Here and in a few other places: Numpy -> NumPy

Could also just omit the NumPy specification or use `np.array` instead perhaps.

ojeda-e commented on membrane_curvature/surface.py:
```diff
@@ -97,11 +97,68 @@ def normalized_grid(grid_z_coordinates, grid_norm_unit):
+    # interpolate values in array
+    interpolated_array = np.interp(index_array,
+                                   np.flatnonzero(~mask_nans),
+                                   array_surface[~mask_nans])
```
I completely missed it and it's a good point. You mean, having this instead, correct?

```
   mask_nans = ~np.isnan(array_surface) #invert original

   index_array = np.arange(array_surface.shape[0])

   interpolated_array = np.interp(index_array,
                                  np.flatnonzero(mask_nans),   #not inverted
                                  array_surface[mask_nans]).   #not inverted
```

ojeda-e commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -208,6 +209,39 @@ def test_get_z_surface(x_bin, y_bin, x_range, y_range, expected_surface):
+])
+def test_surface_interpolation(dummy_surface, expected_interpolated_surface):
+    surface = surface_interpolation(dummy_surface)
+    assert_almost_equal(surface, expected_interpolated_surface)
```
Thanks. I remember @orbeckst [mentioned](https://github.com/MDAnalysis/membrane-curvature/pull/13#discussion_r651379750) that  MDAnalysis typically uses `assert_almost_equal`. Maybe we can have his opinion on what would be best here?

orbeckst replied:
MDAnalysis uses `assert_almost_equal` in preference to `assert_array_almost_equal`. `assert_allclose` might be more recent that when we started using numpy tests so MDAnalysis might just be outdated and should be switching eventually. Follow @tylerjereddy 's advice :-).

ojeda-e commented on membrane_curvature/surface.py:
```diff
@@ -97,11 +97,68 @@ def normalized_grid(grid_z_coordinates, grid_norm_unit):
+
+    """
+
+    interpolated_surface = np.apply_along_axis(interpolation_by_array, axis=0, arr=array_surface)
```
We haven't really assessed performance. I'll skip it for the moment, but will keep it in mind.

Commit:
inverted original mask_nans in interpolation function

Commit:
Replaced assertions by assert_allclose.

Commit:
Added interpolation to base with tests.

ojeda-e commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -132,19 +133,19 @@ def small_grofile():
     H_test = mean_curvature(MEMBRANE_CURVATURE_DATA['z_ref']['small'])
     for h, h_test in zip(MEMBRANE_CURVATURE_DATA['mean_curvature']['small'], H_test):
-        assert_almost_equal(h, h_test)
+        assert_allclose(h, h_test, rtol=6)
```
had to set `rtol=6` here and in the next test otherwise, the tests fail.

lilyminium replied:
Oh dear, that's a very large `rtol`. The default value is 1e-8, for reference. Are you sure you want tests to pass if any calculated value is within 600% of the target value?

lilyminium replied:
If it used to pass with `assert_almost_equal` (where the `decimal=7` by default), you could try setting `atol=1e-7` in `allclose`?

ojeda-e replied:
The test passes with `assert_allclose(h, h_test)` (without specifying `rtol=6`).

ojeda-e commented:
The pushed changes include:
- [x] replaced `assert_almost_equal` by `assert_allclose`. For [two assertions](https://github.com/MDAnalysis/membrane-curvature/pull/52/files#r683844953) I set tolerance to `rtol=6` otherwise the tests fail. Not sure you are ok with this change. (Thanks @tylerjereddy for the suggestion and @orbeckst for confirming)
- [x] added interpolation to `base.py` and tests for grids of different sizes and with a different number of undefined values. 
- [x] In `base.py`, kept the data for `z_surface` and `interpolated_z_surface` as suggested by @lilyminium.

I know `base.py` could be written better, but maybe we can leave that improvement for later and work on the essentials for now if that's ok. 
Hopefully I am not missing relevant things here. Thanks!

ojeda-e commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -132,19 +133,19 @@ def small_grofile():
     K_test = gaussian_curvature(MEMBRANE_CURVATURE_DATA['z_ref']['all'])
     for k, k_test in zip(MEMBRANE_CURVATURE_DATA['gaussian_curvature']['all'], K_test):
-        assert_almost_equal(k, k_test)
+        assert_allclose(k, k_test, rtol=6)
```
It should be `assert_allclose(k, k_test, atol=1e-7)`

ojeda-e commented:
@lilyminium sorry, it's my fault because I made a bit of a mess here.
To clarify. 
`def test_mean_curvature_small` passes with `assert_allclose`.  I don't know why I ended up adding `rtol` it should go.  (I added the comment [here](https://github.com/MDAnalysis/membrane-curvature/pull/52/files#r684656934))

The problem is in `def test_gaussian_curvature_all`. The test passes with `rtol=4` or `atol=1e-7`. From the [docs](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html#numpy.testing.assert_allclose) it says `allclose` "... compares the difference between actual and desired to atol + rtol * abs(desired)."
So probably better have `assert_allclose(k, k_test, atol=1e-7)` ?

Commit:
changed atol in assert_allclose

lilyminium commented:
> So probably better have assert_allclose(k, k_test, atol=1e-7) ?

I think so. Even if you know yourself that the 400% difference is small because the real value is small, it communicates to future users that you expect the degree of difference to be around 0.0000001.

lilyminium commented on membrane_curvature/base.py:
```diff
@@ -179,8 +183,20 @@ def _single_frame(self):
         self.results.average_gaussian = np.nanmean(self.results.gaussian, axis=0)
+        if self.interpolation:
+            self.results.average_interpolated_z_surface = np.nanmean(
+                self.results.interpolated_z_surface[self._frame_index], axis=0)
```
```suggestion
                self.results.interpolated_z_surface, axis=0)
```

And with the others too -- if you specify `frame_index`, you'll wind up only getting the mean of the last frame.

lilyminium replied:
Do you need `np.nanmean`? Do you expect `np.nan` values in your interpolated surfaces? If not, using `np.mean` will reveal errors in case there ever are `np.nan` values.

lilyminium commented on membrane_curvature/surface.py:
```diff
@@ -131,3 +131,60 @@ def normalized_grid(grid_z_coordinates, grid_norm_unit):
+
+def interpolation_by_array(array_surface):
+    """
+        Interpolates values contained in `array_surface` over axis
```
Over which axis?

lilyminium commented on membrane_curvature/surface.py:
```diff
@@ -131,3 +131,60 @@ def normalized_grid(grid_z_coordinates, grid_norm_unit):
+    """
+
+    # create mask for nans
+    mask_nans = ~np.isnan(array_surface)
+
+    # index of array_surface
+    index_array = np.arange(array_surface.shape[0])
+
+    # interpolate values in array
+    interpolated_array = np.interp(index_array,
+                                   np.flatnonzero(mask_nans),
+                                   array_surface[mask_nans])
+
+    return interpolated_array
```
I'm not quite understanding this. `mask_nans` is 2D, with shape (n_x, n_y). `index_array` is 1D, with shape `n_x`. `np.flatnonzero(mask_nans)` and `array_surface[mask_nans]` will be 1D, of length between 0 to n_x * n_y. So I *think* that means that you are only ever operating on the first row of `array_surface`?

I guess my question is then, why is `array_surface` 2D? Why not just pass in one row at a time? In addition, as it is a 2D surface, why use a 1D interpolation function? Why not a [2D one](https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp2d.html)?

lilyminium commented on membrane_curvature/surface.py:
```diff
@@ -131,3 +131,60 @@ def normalized_grid(grid_z_coordinates, grid_norm_unit):
+
+    """
+
+    interpolated_surface = np.apply_along_axis(interpolation_by_array, axis=0, arr=array_surface)
```
Ok, I see that you are applying `interpolation_by_array` by row, then. In that case I suggest amending the documentation of `interpolation_by_array` to specify that the input is 1-dimensional with length `n_x`, instead of `n_x, n_y`. Although I think doing 2D interpolation would be less arbitrary (choosing which axis is x and which is y is basically a random, right?)

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -206,7 +207,38 @@ def test_get_z_surface(x_bin, y_bin, x_range, y_range, expected_surface):
+
+@pytest.mark.parametrize('dummy_surface, expected_interpolated_surface', [
+    # array 3x3 with all 150 and one nan
+    (np.array(([150., 150., 150.],
+               [150., np.nan, 150.],
+               [150., 150., 150.])),
+     np.full((3, 3), 150.)),
+    # array 3x4 with all 150 and two nans
+    (np.array([[150.,  150, 150.,  150.],
+               [150., np.nan, np.nan,  150.],
+               [150., 150., 150.,  150.]]),
+     np.full((3, 4), 150.)),
+    # array 4x4 with all 150 and two nans
+    (np.array([[150., 150,  150.,  150.],
+               [150., np.nan, np.nan,  150.],
+               [150., 130., 140.,  150.],
+               [150., 150., 150.,  150.]]),
+     np.array([[150., 150, 150.,  150.],
+               [150., 140., 145.,  150.],
+               [150., 130., 140.,  150.],
+               [150., 150., 150.,  150.]])),
+    # array 3x3  with lots of nans
+    (np.array([[np.nan, np.nan, 150.],
+              [150, np.nan, 150.],
+              [np.nan, 150., np.nan]]),
+     np.full((3, 3), 150.))
```
Could you please add some tests with more interesting values? e.g. in the one below, the 145 number could have come from interpolating either along the x-axis or the y-axis. In order to be diagnostic of potential current and future bugs, I'd be interested in seeing one where using 1D interpolation gives different results if it's run across the x-axis, vs. the y-axis.

```python
(np.array([[150., 150,  150.,  150.],
               [150., np.nan, np.nan,  150.],
               [150., 130., 140.,  150.],
               [150., 150., 150.,  150.]]),
     np.array([[150., 150, 150.,  150.],
               [150., 140., 145.,  150.],
               [150., 130., 140.,  150.],
               [150., 150., 150.,  150.]])),
```

lilyminium commented on membrane_curvature/tests/test_membrane_curvature.py:
```diff
@@ -563,9 +595,45 @@ def test_analysis_wrapping_coordinates(self, x_bin, y_bin, box_dim, dummy_array,
             MembraneCurvature(universe, wrap=False)
+
+    # test curvature with interpolated surface
+    @pytest.mark.parametrize('dim_x, dim_y, x_bins, y_bins, dummy_array, expected_interp_surf', [
+        # array 3x3 with all 150 and one nan
+        (300, 300, 3, 3, np.array([[0., 0.,   150.], [100., 0.,   150.],   [200., 0.,   150.],
+                                   [0., 100., 150.], [100., 100., np.nan], [200., 100., 150.],
+                                   [0., 200., 150.], [100., 200., 150.],   [200., 200., 150.]]),
+         np.full((1, 3, 3), 150.)),
+        # array 3x3 with all 150 and one nan
+        (300, 300, 3, 3, np.array([[0., 0.,   150.], [100., 0.,   150.],   [200., 0.,   150.],
+                                   [0., 100., 150.], [100., 100., np.nan], [200., 100., 150.],
+                                   [0., 200., np.nan], [100., 200., 150.],   [200., 200., 150.]]),
+         np.full((1, 3, 3), 150.)),
+        # array 3x4 with all 150 and three nans
+        (300, 400, 3, 4, np.array([[0., 0.,   150.], [100., 0.,   150.],   [200., 0.,   150.],
+                                   [0., 100., 150.], [100., 100., np.nan], [200., 100., 150.],
+                                   [0., 200., 150],  [100., 200., np.nan], [200., 200., np.nan],
+                                   [0., 300., 150.], [100., 300., 150.],   [200., 300., 150.]]),
+         np.full((1, 3, 4), 150.)),
+        # array 4x4 with all 120 and many nans
+        (400, 400, 4, 4, np.array([[0., 0., np.nan], [100., 0., 120.], [200., 0., 120.], [300., 0., np.nan],
+                                   [0., 100., 120.], [100., 100., np.nan], [200., 100., 120.], [300., 100., 120.],
+                                   [0., 200., 120], [100., 200., np.nan], [200., 200., np.nan], [300., 200., 120.],
+                                   [0., 300., np.nan], [100., 300., 120.], [200., 300., 120.], [300., 300., np.nan]]),
+         np.full((1, 4, 4), 120.))
+    ])
```
Here as well, tests where the expected values aren't all one number are needed to really check assumptions such as:

* is it interpolating along the axis I think it is
* is it interpolating in a linear way (vs. some random other polynomial)
* is it even interpolating, or just taking the most common number?
* are we sure it's interpolating independently for every frame, or is it interpolating across different frames?

etc.

I am, incidentally, somewhat surprised that coordinates are getting wrapped with `np.nan` values in them. @richardjgowers should this be happening...?

lilyminium reviewed (changes requested):
The tests here are a good start! However I think you definitely need to vary your values (maybe by doing tests on your added data files) for robust checking. I am already reasonably certain that `self.results.average_interpolated_z_surface` etc. are not calculated the way you want them to, which wouldn't have been caught with current test cases (although there aren't actually tests for those yet -- they need some too!).
